File: src/main/java/me/itzisonn_/meazy/addon/datagen/DatagenManager.java
package me.itzisonn_.meazy.addon.datagen;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.JsonDeserializer;
import com.google.gson.JsonSyntaxException;
import com.google.gson.reflect.TypeToken;
import me.itzisonn_.meazy.FileUtils;
import org.jspecify.annotations.NullMarked;

import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.util.HashSet;
import java.util.Set;
import java.util.function.Function;
import java.util.zip.ZipEntry;
import java.util.zip.ZipFile;
import java.util.zip.ZipInputStream;

/**
 * Provides methods for working with datagen
 */
@NullMarked
public class DatagenManager {
    /**
     * Addon's file
     */
    private final File file;

    /**
     * @param file Addon's file
     */
    public DatagenManager(File file) {
        this.file = file;
    }

    /**
     * Gets all lines inside folder with given folderPath and deserializes them using given deserializer. Accepts only single value in JSON
     *
     * @param folderPath Path to datagen folder
     * @param cls Class of deserialized values
     * @param deserializer Json deserializer
     *
     * @return Set of all values inside folder with given folderPath
     * @param <T> Type of deserialized values
     */
    public <T> Set<T> getDeserializedSingle(String folderPath, Class<T> cls, JsonDeserializer<T> deserializer) {
        Set<T> result = new HashSet<>();
        Gson gson = new GsonBuilder().registerTypeAdapter(cls, deserializer).create();

        for (String lines : getDatagenFilesLines(folderPath)) {
            T value = gson.fromJson(lines, cls);
            result.add(value);
        }

        return result;
    }

    /**
     * Gets all lines inside folder with given folderPath and deserializes them using given deserializer. Accepts an array and single value in JSON
     *
     * @param folderPath Path to datagen folder
     * @param cls Class of deserialized values
     * @param deserializer Json deserializer
     *
     * @return Set of all values inside folder with given folderPath
     * @param <T> Type of deserialized values
     */
    @SuppressWarnings("unchecked")
    public <T> Set<T> getDeserializedMultiple(String folderPath, Class<T> cls, JsonDeserializer<T> deserializer) {
        Set<T> result = new HashSet<>();

        Gson gson = new GsonBuilder().registerTypeAdapter(cls, deserializer).create();
        TypeToken<Set<T>> typeToken = (TypeToken<Set<T>>) TypeToken.getParameterized(Set.class, cls);

        for (String lines : getDatagenFilesLines(folderPath)) {
            try {
                T value = gson.fromJson(lines, cls);
                result.add(value);
            }
            catch (JsonSyntaxException e) {
                Set<T> values = gson.fromJson(lines, typeToken);
                result.addAll(values);
            }
        }

        return result;
    }

    /**
     * @param folderPath Path to datagen folder
     * @return Set of all files' lines inside folder with given folderPath
     */
    public Set<String> getDatagenFilesLines(String folderPath) {
        return new HashSet<>(getDatagenInputStreams(folderPath, FileUtils::getLines));
    }

    private <T> Set<T> getDatagenInputStreams(String folderPath, Function<InputStream, T> converter) {
        Set<T> result = new HashSet<>();

        try (ZipFile zipFile = new ZipFile(file)) {
            ZipInputStream inputStream = new ZipInputStream(new FileInputStream(file));

            ZipEntry zipEntry = inputStream.getNextEntry();
            while (zipEntry != null) {
                if (!zipEntry.getName().startsWith("data/" + folderPath + "/") || zipEntry.isDirectory()) {
                    zipEntry = inputStream.getNextEntry();
                    continue;
                }

                result.add(converter.apply(zipFile.getInputStream(zipEntry)));
                zipEntry = inputStream.getNextEntry();
            }

            return result;
        }
        catch (IOException e) {
            throw new RuntimeException(e);
        }
    }
}

File: src/main/java/me/itzisonn_/meazy/datagen/DatagenManager.kt
package me.itzisonn_.meazy.datagen

import com.google.gson.GsonBuilder
import com.google.gson.JsonDeserializer
import com.google.gson.JsonSyntaxException
import com.google.gson.reflect.TypeToken
import me.itzisonn_.meazy.util.FileUtils
import java.io.IOException
import java.net.URISyntaxException
import java.nio.file.Files
import java.nio.file.Path
import kotlin.io.path.toPath
import kotlin.io.path.walk

/**
 * Provides methods for working with datagen
 */
class DatagenManager {
    /**
     * Gets all lines inside folder with given folderPath and deserializes them using given deserializer. Accepts only single value in JSON
     *
     * @param folderPath Path to datagen folder
     * @param cls Class of deserialized values
     * @param deserializer Json deserializer
     *
     * @return Set of all values inside folder with given folderPath
     * @param T Type of deserialized values
     */
    fun <T> getDeserializedSingle(folderPath: String, cls: Class<T>, deserializer: JsonDeserializer<T>): Set<T> {
        val result = mutableSetOf<T>()
        val gson = GsonBuilder().registerTypeAdapter(cls, deserializer).create()

        for (lines in getDatagenFilesLines(folderPath)) {
            val value = gson.fromJson(lines, cls)
            result.add(value)
        }

        return result
    }

    /**
     * Gets all lines inside folder with given folderPath and deserializes them using given deserializer. Accepts an array and single value in JSON
     *
     * @param folderPath Path to datagen folder
     * @param cls Class of deserialized values
     * @param deserializer Json deserializer
     *
     * @return Set of all values inside folder with given folderPath
     * @param T Type of deserialized values
     */
    fun <T> getDeserializedMultiple(folderPath: String, cls: Class<T>, deserializer: JsonDeserializer<T>): Set<T> {
        val result = mutableSetOf<T>()

        val gson = GsonBuilder().registerTypeAdapter(cls, deserializer).create()
        @Suppress("UNCHECKED_CAST")
        val typeToken = TypeToken.getParameterized(MutableSet::class.java, cls) as TypeToken<MutableSet<T>>

        for (lines in getDatagenFilesLines(folderPath)) {
            try {
                val value = gson.fromJson(lines, cls)
                result.add(value)
            }
            catch (_: JsonSyntaxException) {
                val values = gson.fromJson(lines, typeToken)
                result.addAll(values)
            }
        }

        return result
    }

    /**
     * @param folderPath Path to datagen folder
     * @return Set of all files' lines inside folder with given folderPath
     */
    fun getDatagenFilesLines(folderPath: String): Set<String> {
        val url = this::class.java.getResource("/data/$folderPath") ?: error("Can't find file: $folderPath")
        val path = url.toURI().toPath()

        try {
            return path.walk()
                .filter { path: Path -> Files.isRegularFile(path) }
                .map { obj: Path -> FileUtils.getLines(obj) }
                .toSet()
        }
        catch (e: IOException) {
            throw RuntimeException(e)
        }
        catch (e: URISyntaxException) {
            throw RuntimeException(e)
        }
    }
}